pkg/controllers/updaterun/stop.go

```diff
@@ -0,0 +1,237 @@
+	for _, clusterStatus := range existingDeleteStageClusterMap {
+		// Make sure the cluster is marked as deleted.
+		if !condition.IsConditionStatusTrue(meta.FindStatusCondition(clusterStatus.Conditions, string(placementv1beta1.ClusterUpdatingConditionStarted)), updateRun.GetGeneration()) {
+			markClusterUpdatingStarted(clusterStatus, updateRun.GetGeneration())
+		}
+		markClusterUpdatingSucceeded(clusterStatus, updateRun.GetGeneration())
+	}
+
+	klog.V(2).InfoS("The delete stage is stopping", "numberOfDeletingClusters", len(toBeDeletedBindings), "updateRun", updateRunRef)
+	if len(toBeDeletedBindings) == 0 {
+		markStageUpdatingStopped(updateRunStatus.DeletionStageStatus, updateRun.GetGeneration())
+	}
+	return len(toBeDeletedBindings) == 0, nil
+}
+
+// markUpdateRunStopping marks the update run as stopping in memory.
+func markUpdateRunStopping(updateRun placementv1beta1.UpdateRunObj) {
+	klog.V(2).InfoS("Marking the update run as stopping", "updateRun", klog.KObj(updateRun))
+	updateRunStatus := updateRun.GetUpdateRunStatus()
+	meta.SetStatusCondition(&updateRunStatus.Conditions, metav1.Condition{
+		Type:               string(placementv1beta1.StagedUpdateRunConditionProgressing),
+		Status:             metav1.ConditionUnknown,
+		ObservedGeneration: updateRun.GetGeneration(),
+		Reason:             condition.UpdateRunStoppingReason,
+		Message:            "The update run is the process of stopping, waiting for all the updating/deleting clusters to finish updating before completing the stop process",
+	})
+}
+
+// markStageUpdatingStopping marks the stage updating status as pausing in memory.
+func markStageUpdatingStopping(stageUpdatingStatus *placementv1beta1.StageUpdatingStatus, generation int64) {
+	meta.SetStatusCondition(&stageUpdatingStatus.Conditions, metav1.Condition{
+		Type:               string(placementv1beta1.StageUpdatingConditionProgressing),
+		Status:             metav1.ConditionUnknown,
+		ObservedGeneration: generation,
+		Reason:             condition.StageUpdatingStoppingReason,
+		Message:            "Waiting for all the updating clusters to finish updating before completing the stop process",
+	})
+}
+
+// markStageUpdatingStopped marks the stage updating status as stopped in memory.
+func markStageUpdatingStopped(stageUpdatingStatus *placementv1beta1.StageUpdatingStatus, generation int64) {
+	meta.SetStatusCondition(&stageUpdatingStatus.Conditions, metav1.Condition{
+		Type:               string(placementv1beta1.StageUpdatingConditionProgressing),
+		Status:             metav1.ConditionFalse,
+		ObservedGeneration: generation,
+		Reason:             condition.StageUpdatingStoppedReason,
+		Message:            "All the updating clusters have finished updating, the stage is now stopped, waiting to be resumed",
+	})
+}
+
+func checkIfErrorStagedUpdateAborted(err error, updateRun placementv1beta1.UpdateRunObj, updatingStageStatus *placementv1beta1.StageUpdatingStatus) {
```

Comment:
nit: we should move this to a common util file since it's used across execution, stop states